In [1]:
datafile = "nyc_311_data_subset-2.csv"

In [2]:
import pandas as pd
import numpy as np

<h4><b>read_csv</b>: A pandas function that reads a comma separated file</h4>
read_csv will try to format the data so that it is the correct type and will report any typing problems<br>
It will also look for a header row. 
<br>http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html

In [3]:
data = pd.read_csv(datafile)
data

/Users/Beck/anaconda/envs/tflearn/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unique Key    Created Date     Closed Date Agency Incident Zip  \
0                1  10/11/16 11:53  10/11/16 12:00   DSNY          NaN   
1                2  10/11/16 11:36  10/11/16 12:00   DSNY          NaN   
2                3  10/11/16 11:36  10/11/16 12:00   DSNY          NaN   
3                4  10/11/16 12:39  10/11/16 12:39   DSNY          NaN   
4                5  10/11/16 12:18  10/11/16 12:18   DSNY          NaN   
5                6  10/11/16 11:36  10/11/16 12:00   DSNY          NaN   
6                7  10/11/16 12:39  10/11/16 12:39   DSNY          NaN   
7                8  10/11/16 11:44             NaN   DSNY          NaN   
8                9  10/11/16 12:35  10/11/16 12:35   DSNY          NaN   
9               10  10/11/16 11:26   5/27/16 12:00   DSNY          NaN   
10              11     9/1/16 0:33    9/16/16 1:06    DCA        10001   
11              12    9/1/16 20:16   9/10/16 18:08    HPD        11691   
12              13    9/1/16 12:17    9/7/16 12:00   DSNY        11211   
13              14    9/1/16 12:10   9/10/16 14:23    HPD        10027   
14              15    9/1/16 12:32    9/11/16 2:03    HPD        10452   
15              16    9/1/16 20:16   9/10/16 18:08    HPD        11691   
16              17     9/1/16 8:35    9/7/16 12:00   DSNY        11428   
17              18    9/1/16 13:19   9/16/16 14:32    DOT        11101   
18              19    9/1/16 11:00    9/8/16 12:00   DSNY        10075   
19              20    9/1/16 11:45    9/3/16 12:00   DSNY        11215   
20              21    9/1/16 10:11   9/16/16 11:11    DOT        11101   
21              22     9/1/16 8:22    9/16/16 7:42    DOT        11210   
22              23    9/1/16 17:31   9/16/16 10:10    DOT        11231   
23              24     9/1/16 8:50   9/16/16 11:12    DOT        11210   
24              25    9/1/16 14:19   9/16/16 14:11    DOT        11101   
25              26    9/1/16 12:46   9/16/16 11:53    DOT        11217   
26              27    9/1/16 13:33   9/16/16 12:55    DOT        11231   
27              28    9/1/16 13:51    9/11/16 2:05    HPD        10457   
28              29    9/1/16 12:28    9/1/16 12:28    DOT          NaN   
29              30    9/1/16 12:34   9/10/16 15:01    HPD        10033   
...            ...             ...             ...    ...          ...   
971033      971034   9/20/16 13:36   9/20/16 16:14   NYPD        11357   
971034      971035   9/20/16 21:22   9/21/16 10:21    DOT        11368   
971035      971036   9/20/16 12:27    9/21/16 0:30    DOT        10461   
971036      971037   9/20/16 14:25   9/26/16 13:57    DOF          NaN   
971037      971038   9/20/16 13:55    9/22/16 8:10    DEP        11217   
971038      971039   9/20/16 13:25   9/21/16 10:59    DOF        10013   
971039      971040   9/20/16 14:08   9/21/16 14:08    DEP        11238   
971040      971041    9/20/16 7:51   9/21/16 11:46    DOT        11222   
971041      971042   9/20/16 19:57   9/20/16 19:58    HRA          NaN   
971042      971043    9/20/16 8:56   9/20/16 10:20    DEP        11209   
971043      971044    9/20/16 6:22   9/23/16 12:45    DEP        10029   
971044      971045    9/20/16 3:40             NaN    DEP        11366   
971045      971046    9/20/16 0:06   9/22/16 22:45    DEP        10027   
971046      971047    9/20/16 3:05    9/20/16 7:58   NYPD        10030   
971047      971048    9/20/16 9:19   9/27/16 15:20    DOT          NaN   
971048      971049   9/20/16 20:42    1/13/17 8:56    DOT        10314   
971049      971050   9/20/16 16:42   9/20/16 18:40    DOT          NaN   
971050      971051   9/20/16 11:47   9/16/16 11:47    DOT        11414   
971051      971052   9/20/16 18:31   9/20/16 20:42   NYPD        11205   
971052      971053   9/20/16 19:35   9/20/16 21:39   NYPD        11205   
971053      971054    9/20/16 8:10   9/23/16 11:28    DOT          NaN   
971054      971055   9/20/16 10:19    9/21/16 5:09    DHS        10025   
971055      

<h4>Let's examine our data</h4>

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 971063 entries, 0 to 971062
Data columns (total 8 columns):
Unique Key      971063 non-null int64
Created Date    971063 non-null object
Closed Date     882944 non-null object
Agency          971063 non-null object
Incident Zip    911140 non-null object
Borough         971063 non-null object
Latitude        887284 non-null float64
Longitude       887284 non-null float64
dtypes: float64(2), int64(1), object(5)
memory usage: 59.3+ MB


<h4>Looks like Unique Key really is a unique key and can serve as an index</h4>

In [5]:
data = pd.read_csv(datafile,index_col='Unique Key')

/Users/Beck/anaconda/envs/tflearn/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [6]:
data.iloc[1:10]

Created Date     Closed Date Agency Incident Zip Borough  \
Unique Key                                                               
2           10/11/16 11:36  10/11/16 12:00   DSNY          NaN  QUEENS   
3           10/11/16 11:36  10/11/16 12:00   DSNY          NaN  QUEENS   
4           10/11/16 12:39  10/11/16 12:39   DSNY          NaN  QUEENS   
5           10/11/16 12:18  10/11/16 12:18   DSNY          NaN  QUEENS   
6           10/11/16 11:36  10/11/16 12:00   DSNY          NaN  QUEENS   
7           10/11/16 12:39  10/11/16 12:39   DSNY          NaN  QUEENS   
8           10/11/16 11:44             NaN   DSNY          NaN  QUEENS   
9           10/11/16 12:35  10/11/16 12:35   DSNY          NaN  QUEENS   
10          10/11/16 11:26   5/27/16 12:00   DSNY          NaN  QUEENS   

            Latitude  Longitude  
Unique Key                       
2                NaN        NaN  
3                NaN        NaN  
4                NaN        NaN  
5                NaN        NaN  
6                NaN        NaN  
7                NaN        NaN  
8                NaN        NaN  
9                NaN        NaN  
10               NaN        NaN

<h4>Columns 4 has mixed types</h4>

<h4>Column 4 is incident zip</h4>
Let's examine it<br>
The unique() function returns unique values in a column

In [7]:
data['Incident Zip'].unique()

array([nan, '10001', '11691', '11211', '10027', '10452', '11428', '11101',
       '10075', '11215', '11210', '11231', '11217', '10457', '10033',
       '11209', '11201', '11367', '10029', '10021', '10028', '10034',
       '10032', '10039', '11414', '10461', '11229', '10462', '11223',
       '10023', '10453', '11225', '11219', '10451', '11234', '10014',
       '11354', '11361', '10468', '11233', '10466', '11204', '11413',
       '11224', '11375', '11040', '11232', '11203', '11205', '11434',
       '10011', '10003', '10025', '10013', '10036', '11237', '11355',
       '11368', '10454', '10456', '10463', '11222', '11228', '11216',
       '10128', '11435', '11419', '11358', '11421', '10019', '11238',
       '11213', '11235', '11420', '10038', '11226', '10472', '10016',
       '11221', '11236', '11436', '11214', '11377', '11385', '11365',
       '10312', '11426', '11373', '11218', '10005', '11230', '10026',
       '10473', '10280', '10301', '10309', '10310', '10009', '10002',
       '11433',

<h4>Some issues</h4>
<li>Sometimes zip is a float, other times it is a str
<li>Zipcodes that are represented as floats and start with 0 are missing the first digit
<li>Some zipcodes have the 4 digit extension added. Comparison becomes tough
<li>What the heck is zip 0?
<li>What about the missing (nan) values? The ? (question mark)? "UNKNOWN"?

<h2>The first step in data cleaning is to:</h2>
<h3>Decide what to do with "bad" data ("JFK", "UNKNOWN", etc.). Convert to Nan or delete the record.</h3>
<h3>Make sure all data in a column is in the correct format (convert floats to strings, get rid of the 4 digit extension)</h3>
<h3>Decide what to do with missing values (NaNs)</h3>

<h3>for "Incident Zip"</h3>
<h4>we'll drop rows with NaN or bad data</h4>
<h4>get rid of the 4 digit extension</h4>
<h4>remove zips less than 10000 and greater than 19999</h4>
<h3>Let's write a function that fixes zips</h3>

In [8]:
def fix_zip(input_zip):
    try:
        input_zip = int(float(input_zip))
    except:
        try:
            input_zip = int(input_zip.split('-')[0])
        except:
            return np.NaN
    if input_zip < 10000 or input_zip > 19999:
        return np.NaN
    return str(input_zip)
        

<h4>And test it</h4>

In [9]:
fix_zip('11211.00')

'11211'

<h3>Next, we'll apply this function to every element in input zip to get a revised column</h3>
<h4>The pandas function "apply" applies a function to a dataframe column
<li>fix_zip will be applied to each element of the Incident Zip column and we replace the existing column with the modified one

In [10]:
data['Incident Zip'] = data['Incident Zip'].apply(fix_zip)

In [11]:
data['Incident Zip'].unique()

array([nan, '10001', '11691', '11211', '10027', '10452', '11428', '11101',
       '10075', '11215', '11210', '11231', '11217', '10457', '10033',
       '11209', '11201', '11367', '10029', '10021', '10028', '10034',
       '10032', '10039', '11414', '10461', '11229', '10462', '11223',
       '10023', '10453', '11225', '11219', '10451', '11234', '10014',
       '11354', '11361', '10468', '11233', '10466', '11204', '11413',
       '11224', '11375', '11040', '11232', '11203', '11205', '11434',
       '10011', '10003', '10025', '10013', '10036', '11237', '11355',
       '11368', '10454', '10456', '10463', '11222', '11228', '11216',
       '10128', '11435', '11419', '11358', '11421', '10019', '11238',
       '11213', '11235', '11420', '10038', '11226', '10472', '10016',
       '11221', '11236', '11436', '11214', '11377', '11385', '11365',
       '10312', '11426', '11373', '11218', '10005', '11230', '10026',
       '10473', '10280', '10301', '10309', '10310', '10009', '10002',
       '11433',

<h3>Finally, we'll get rid of all rows that have zip == Nan</h3>
<li>We don't have to, that's just a choice we're making</li>

In [12]:
data = data[data['Incident Zip'].notnull()]

In [13]:
data

Created Date     Closed Date Agency Incident Zip        Borough  \
Unique Key                                                                     
11            9/1/16 0:33    9/16/16 1:06    DCA        10001      MANHATTAN   
12           9/1/16 20:16   9/10/16 18:08    HPD        11691         QUEENS   
13           9/1/16 12:17    9/7/16 12:00   DSNY        11211       BROOKLYN   
14           9/1/16 12:10   9/10/16 14:23    HPD        10027      MANHATTAN   
15           9/1/16 12:32    9/11/16 2:03    HPD        10452          BRONX   
16           9/1/16 20:16   9/10/16 18:08    HPD        11691         QUEENS   
17            9/1/16 8:35    9/7/16 12:00   DSNY        11428         QUEENS   
18           9/1/16 13:19   9/16/16 14:32    DOT        11101         QUEENS   
19           9/1/16 11:00    9/8/16 12:00   DSNY        10075      MANHATTAN   
20           9/1/16 11:45    9/3/16 12:00   DSNY        11215       BROOKLYN   
21           9/1/16 10:11   9/16/16 11:11    DOT        11101         QUEENS   
22            9/1/16 8:22    9/16/16 7:42    DOT        11210       BROOKLYN   
23           9/1/16 17:31   9/16/16 10:10    DOT        11231       BROOKLYN   
24            9/1/16 8:50   9/16/16 11:12    DOT        11210       BROOKLYN   
25           9/1/16 14:19   9/16/16 14:11    DOT        11101         QUEENS   
26           9/1/16 12:46   9/16/16 11:53    DOT        11217       BROOKLYN   
27           9/1/16 13:33   9/16/16 12:55    DOT        11231       BROOKLYN   
28           9/1/16 13:51    9/11/16 2:05    HPD        10457          BRONX   
30           9/1/16 12:34   9/10/16 15:01    HPD        10033      MANHATTAN   
31            9/1/16 7:03     9/7/16 7:30    DEP        11209       BROOKLYN   
32           9/1/16 12:29    9/11/16 2:03    HPD        10452          BRONX   
33           9/1/16 17:32    9/16/16 7:20    DOT        11201       BROOKLYN   
34            9/1/16 8:53    9/16/16 9:03    DOT        11201       BROOKLYN   
35           9/1/16 17:36    9/16/16 9:47    DOT        11210       BROOKLYN   
36           9/1/16 17:46    9/16/16 9:49    DOT        11367         QUEENS   
37           9/1/16 13:12    9/2/16 12:00   DSNY        10029      MANHATTAN   
38            9/1/16 9:05    9/8/16 12:00   DSNY        10021      MANHATTAN   
39           9/1/16 12:00   9/16/16 11:48    DOT        10028      MANHATTAN   
40           9/1/16 12:19    9/16/16 7:20    DOT        10034      MANHATTAN   
41            9/1/16 9:57    9/16/16 9:01    DOT        11217       BROOKLYN   
...                   ...             ...    ...          ...            ...   
971027      9/20/16 10:35   10/6/16 11:41    DOF        10019      MANHATTAN   
971028       9/20/16 9:14   9/20/16 12:53    DOF        11201       BROOKLYN   
971030      9/20/16 16:27    9/21/16 9:38    DOF        10466          BRONX   
971031      9/20/16 16:13   9/21/16 11:00    DOT        10462          BRONX   
971032      9/20/16 13:08   9/28/16 21:55    DOT        10468          BRONX   
971034      9/20/16 13:36   9/20/16 16:14   NYPD        11357         QUEENS   
971035      9/20/16 21:22   9/21/16 10:21    DOT        11368         QUEENS   
971036      9/20/16 12:27    9/21/16 0:30    DOT        10461          BRONX   
971038      9/20/16 13:55    9/22/16 8:10    DEP        11217       BROOKLYN   
971039      9/20/16 13:25   9/21/16 10:59    DOF        10013      MANHATTAN   
971040      9/20/16 14:08   9/21/16 14:08    DEP        11238       BROOKLYN   
971041       9/20/16 7:51   9/21/16 11:46    DOT        11222       BROOKLYN   
971043       9/20/16 8:56   9/20/16 10:20    DEP        11209       BROOKLYN   
971044       9/20/16 6:22   9/23/16 12:45    DEP        10029      MANHATTAN   
971045       9/20/16 3:40             NaN    DEP        11366         QUEENS   
971046       9/20/16 0:06   9/22/16 22:45    DEP        10027      MANHATTAN   
971047       9/20/16 3:05    9/20/16 7:58   NYPD        10030      MANHATTAN   
971049      9

<h3>Let's take a look at the columns again</h3>

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 910907 entries, 11 to 971063
Data columns (total 7 columns):
Created Date    910907 non-null object
Closed Date     829453 non-null object
Agency          910907 non-null object
Incident Zip    910907 non-null object
Borough         910907 non-null object
Latitude        887168 non-null float64
Longitude       887168 non-null float64
dtypes: float64(2), object(5)
memory usage: 55.6+ MB


<h3>Closed Data, Latitude and Longitude all have missing values</h3>
<h3>Let's get rid of them</h3>

In [15]:
data = data[(data['Latitude'].notnull()) & (data['Longitude'].notnull())  & (data['Closed Date'].notnull())]

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 806561 entries, 11 to 971063
Data columns (total 7 columns):
Created Date    806561 non-null object
Closed Date     806561 non-null object
Agency          806561 non-null object
Incident Zip    806561 non-null object
Borough         806561 non-null object
Latitude        806561 non-null float64
Longitude       806561 non-null float64
dtypes: float64(2), object(5)
memory usage: 49.2+ MB


<h4>Let's take a look at Borough data</h4>

In [17]:
data['Borough'].unique()

array(['MANHATTAN', 'QUEENS', 'BROOKLYN', 'BRONX', 'STATEN ISLAND',
       'Unspecified'], dtype=object)

<h4>Let's look at 'Unspecified'</h4>

In [18]:
data[data['Borough']=='Unspecified'][['Agency','Incident Zip']]

Agency Incident Zip
Unique Key                    
28854        NYPD        10312
29525        NYPD        11368
29559        NYPD        11422
29607        NYPD        10454
48565        NYPD        11209
48572        NYPD        11226
48657        NYPD        11225
48743        NYPD        11373
48781        NYPD        10467
48940         DPR        10454
49258        NYPD        11204
49264        NYPD        10039
49299        NYPD        11369
49471        NYPD        10454
49700        NYPD        11217
49919        NYPD        11367
50081        NYPD        11217
50542        NYPD        11234
50865        NYPD        10452
51202        NYPD        11364
72977        NYPD        10014
73118        NYPD        11385
73119        NYPD        10453
73181        NYPD        11101
73241        NYPD        11203
73286        NYPD        11432
73323        NYPD        11234
73433        NYPD        10461
73434        NYPD        11210
73499        NYPD        10314
...           ...          ...
188498       NYPD        10028
188500       NYPD        11217
188505       NYPD        10040
188506       NYPD        10029
188507       NYPD        10014
188521       NYPD        11208
188536        DHS        10023
188542        TLC        10023
188543        DHS        10023
188544        TLC        11230
188545        DHS        10017
188553        DHS        10024
188558        DHS        10016
188597       NYPD        11206
188599       NYPD        11373
188627       NYPD        11417
188628       NYPD        11418
188633       NYPD        10023
188634       NYPD        10472
188636       NYPD        11236
188639       NYPD        11366
188640       NYPD        10001
188642       NYPD        11237
188647       NYPD        10312
188648       NYPD        11223
188649       NYPD        11373
188650       NYPD        10468
188651       NYPD        11218
214971       NYPD        11233
235626       NYPD        11249

[819 rows x 2 columns]

<h4>Looks like a lot of these are NYPD related</h4>
<h4>Let's take a closer look</h4>

In [19]:
data[data['Borough']=='Unspecified'].groupby('Agency').count()

Created Date  Closed Date  Incident Zip  Borough  Latitude  Longitude
Agency                                                                       
3/1/01             1            1             1        1         1          1
DHS               67           67            67       67        67         67
DOE                1            1             1        1         1          1
DOF                3            3             3        3         3          3
DOT               13           13            13       13        13         13
DPR                2            2             2        2         2          2
FDNY               1            1             1        1         1          1
NYPD             725          725           725      725       725        725
TLC                6            6             6        6         6          6

<h4>Unspecified appears to have a systematic bias toward NYPD</h4>
<h4>Though only a small proportion of NYPD complaints (see below)</h4>
<h4>We have to decide whether to keep them or lose them!</h4>

In [20]:
nypd_complaints_total = data[data['Agency']=='NYPD']['Borough'].count()
nypd_unspecified = data[(data['Borough']=='Unspecified') & (data['Agency']=="NYPD")]['Borough'].count()
percentage = nypd_unspecified/nypd_complaints_total*100
print("%1.2f"%percentage)

0.26


<h3>For now, we'll get rid of them. Unspecified will be hard to explain!</h3>

In [21]:
data = data[data['Borough'] != 'Unspecified']

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 805742 entries, 11 to 971063
Data columns (total 7 columns):
Created Date    805742 non-null object
Closed Date     805742 non-null object
Agency          805742 non-null object
Incident Zip    805742 non-null object
Borough         805742 non-null object
Latitude        805742 non-null float64
Longitude       805742 non-null float64
dtypes: float64(2), object(5)
memory usage: 49.2+ MB


<h4>Dealing with time</h4>
<li>Dates and times are best converted to datetime
<li>That way they will be useful for analysis because we can compute timedelta objects

In [23]:
data['Created Date']

Unique Key
11          9/1/16 0:33
12         9/1/16 20:16
13         9/1/16 12:17
14         9/1/16 12:10
15         9/1/16 12:32
16         9/1/16 20:16
17          9/1/16 8:35
18         9/1/16 13:19
19         9/1/16 11:00
20         9/1/16 11:45
21         9/1/16 10:11
22          9/1/16 8:22
23         9/1/16 17:31
24          9/1/16 8:50
25         9/1/16 14:19
26         9/1/16 12:46
27         9/1/16 13:33
28         9/1/16 13:51
30         9/1/16 12:34
31          9/1/16 7:03
32         9/1/16 12:29
33         9/1/16 17:32
34          9/1/16 8:53
35         9/1/16 17:36
36         9/1/16 17:46
37         9/1/16 13:12
38          9/1/16 9:05
39         9/1/16 12:00
40         9/1/16 12:19
41          9/1/16 9:57
              ...      
971014    9/20/16 20:01
971018    9/20/16 22:14
971021    9/18/16 11:07
971022    9/20/16 11:02
971023    9/19/16 16:31
971025    9/20/16 19:09
971031    9/20/16 16:13
971032    9/20/16 13:08
971034    9/20/16 13:36
971035    9/20/16 21:22
97103

In [24]:
import datetime
#data['Created Date'] = data['Created Date'].apply(lambda x:datetime.datetime.strptime(x,'%m/%d/%Y %I:%M:%S %p'))
data['Created Date'] = data['Created Date'].apply(lambda x:datetime.datetime.strptime(x,'%m/%d/%y %H:%M'))

In [25]:
data['Created Date'][0:20]

Unique Key
11   2016-09-01 00:33:00
12   2016-09-01 20:16:00
13   2016-09-01 12:17:00
14   2016-09-01 12:10:00
15   2016-09-01 12:32:00
16   2016-09-01 20:16:00
17   2016-09-01 08:35:00
18   2016-09-01 13:19:00
19   2016-09-01 11:00:00
20   2016-09-01 11:45:00
21   2016-09-01 10:11:00
22   2016-09-01 08:22:00
23   2016-09-01 17:31:00
24   2016-09-01 08:50:00
25   2016-09-01 14:19:00
26   2016-09-01 12:46:00
27   2016-09-01 13:33:00
28   2016-09-01 13:51:00
30   2016-09-01 12:34:00
31   2016-09-01 07:03:00
Name: Created Date, dtype: datetime64[ns]

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 805742 entries, 11 to 971063
Data columns (total 7 columns):
Created Date    805742 non-null datetime64[ns]
Closed Date     805742 non-null object
Agency          805742 non-null object
Incident Zip    805742 non-null object
Borough         805742 non-null object
Latitude        805742 non-null float64
Longitude       805742 non-null float64
dtypes: datetime64[ns](1), float64(2), object(4)
memory usage: 69.2+ MB


In [27]:
data['Closed Date'] = data['Closed Date'].apply(lambda x:datetime.datetime.strptime(x,'%m/%d/%y %H:%M'))

In [28]:
data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 805742 entries, 11 to 971063
Data columns (total 7 columns):
Created Date    805742 non-null datetime64[ns]
Closed Date     805742 non-null datetime64[ns]
Agency          805742 non-null object
Incident Zip    805742 non-null object
Borough         805742 non-null object
Latitude        805742 non-null float64
Longitude       805742 non-null float64
dtypes: datetime64[ns](2), float64(2), object(3)
memory usage: 69.2+ MB


<h3>We can create a new column that tracks the time it takes to close a complaint</h3>

In [29]:
# pandas 需要再0.20.3才可以. 0.19.x 是会报错.
data['processing_time'] =  data['Closed Date'] - data['Created Date']

In [30]:
#And look at summary statistics
data['processing_time'].describe()


count                     805742
mean      5 days 00:05:12.826909
std      12 days 06:08:18.124834
min          -134 days +00:00:00
25%              0 days 02:35:00
50%              0 days 21:11:00
75%              4 days 14:30:00
max            148 days 13:11:00
Name: processing_time, dtype: object

<h4>There is some odd stuff here</h4>
<li>Negative processing time?
<li>Since our data is for two months, a max of 148 days worth checking out

<h3>Let's examine the negative processing time data</h3>

In [31]:
data[data['processing_time']<datetime.timedelta(0,0,0)]

Created Date         Closed Date Agency Incident Zip  \
Unique Key                                                               
3972       2016-09-16 14:24:00 2016-09-15 14:23:00    DOT        10314   
5223       2016-09-20 14:03:00 2016-09-16 14:03:00    DOT        11220   
6084       2016-10-20 11:24:00 2016-10-19 01:24:00    DOT        11412   
6085       2016-10-20 16:40:00 2016-10-19 16:39:00    DOT        10306   
6086       2016-10-20 12:19:00 2016-10-19 12:18:00    DOT        11209   
6115       2016-10-20 12:05:00 2016-10-19 02:05:00    DOT        11208   
6396       2016-10-24 10:37:00 2016-10-21 10:37:00    DOT        11691   
6682       2016-10-31 10:26:00 2016-10-28 10:26:00    DOT        11417   
6694       2016-10-31 10:46:00 2016-10-27 10:46:00    DOT        11362   
6819       2016-09-20 11:49:00 2016-09-16 11:49:00    DOT        11432   
6820       2016-09-20 14:16:00 2016-09-16 14:16:00    DOT        11238   
6830       2016-09-20 14:34:00 2016-09-16 14:34:00    DOT        11201   
6836       2016-09-20 14:04:00 2016-09-16 14:00:00    DOT        11220   
6840       2016-09-20 12:00:00 2016-09-16 12:00:00    DOT        11434   
6841       2016-09-20 11:48:00 2016-09-16 11:48:00    DOT        11432   
6842       2016-09-20 11:43:00 2016-09-16 11:43:00    DOT        11416   
6843       2016-09-20 14:23:00 2016-09-16 14:23:00    DOT        11221   
6845       2016-09-20 11:48:00 2016-09-16 11:48:00    DOT        11432   
6866       2016-09-20 14:03:00 2016-09-16 14:03:00    DOT        11219   
6867       2016-09-20 11:59:00 2016-09-16 11:59:00    DOT        11433   
6868       2016-09-20 11:54:00 2016-09-16 11:54:00    DOT        11374   
6869       2016-09-20 12:53:00 2016-09-16 12:53:00    DOT        11224   
6877       2016-09-20 14:51:00 2016-09-19 14:50:00    DOT        11225   
6885       2016-09-20 12:00:00 2016-09-16 12:00:00    DOT        11412   
6890       2016-09-20 11:49:00 2016-09-16 11:49:00    DOT        11432   
6891       2016-09-20 11:41:00 2016-09-16 11:41:00    DOT        11105   
6893       2016-09-20 12:50:00 2016-09-16 12:50:00    DOT        11212   
6895       2016-09-20 14:32:00 2016-09-16 14:32:00    DOT        11217   
6896       2016-09-20 12:54:00 2016-09-16 12:54:00    DOT        11214   
6899       2016-09-20 12:31:00 2016-09-17 22:40:00    DOT        10469   
...                        ...                 ...    ...          ...   
957010     2016-10-17 00:00:00 2016-09-12 00:00:00  DOHMH        10025   
957015     2016-10-17 00:00:00 2016-09-13 00:00:00  DOHMH        10037   
957080     2016-10-17 00:00:00 2016-10-11 00:00:00  DOHMH        10461   
957095     2016-10-17 00:00:00 2016-10-03 00:00:00  DOHMH        10026   
957167     2016-10-17 00:00:00 2016-10-03 00:00:00  DOHMH        11385   
957189     2016-10-17 00:00:00 2016-10-05 00:00:00  DOHMH        10027   
958744     2016-09-19 16:54:00 2016-09-17 00:26:00    DOT        10303   
958745     2016-09-19 12:18:00 2016-09-17 00:15:00    DOT        10465   
958746     2016-09-19 12:05:00 2016-09-16 22:52:00    DOT        10460   
958747     2016-09-19 16:53:00 2016-09-16 22:57:00    DOT        10306   
958748     2016-09-19 11:52:00 2016-09-16 11:52:00    DOT        10474   
958752     2016-09-20 12:01:00 2016-09-16 12:01:00    DOT        11434   
958753     2016-09-20 14:12:00 2016-09-16 14:12:00    DOT        11226   
958758     2016-09-20 11:53:00 2016-09-16 11:53:00    DOT        11694   
958759     2016-09-20 14:33:00 2016-09-16 14:33:00    DOT        11201   
958760     2016-09-20 14:31:00 2016-09-16 14:31:00    DOT        11215   
958804     2016-09-22 12:15:00 2016-09-21 12:14:00    DOT        11233   
958858     2016-09-26 16:23:00 2016-09-23 23:00:00    DOT        10308   
958859     2016-09-26 13:31:00 2016-09-23 21:20:00    DOT        10474   
958860     2016-09-26 12:39:00 2016-09-23 12:39:00    DOT        11232   
958861     2016-09-26 12:28:00 2016-09-23 12:28:00    DOT        11211   
965521     2016-09-13 

<h3>And the large processing times as well</h3>

In [34]:
data[data['processing_time']>datetime.timedelta(140,0,0)]

Created Date         Closed Date Agency Incident Zip  \
Unique Key                                                               
5371       2016-09-01 12:00:00 2017-01-24 00:00:00    DOB        11105   
28919      2016-09-02 21:11:00 2017-01-27 00:00:00    DOB        11385   
155601     2016-09-01 19:53:00 2017-01-27 11:59:00    DPR        11417   
155747     2016-09-01 19:24:00 2017-01-24 00:00:00    DOB        11385   
155818     2016-09-01 10:49:00 2017-01-28 00:00:00    DOB        11691   
156081     2016-09-01 19:34:00 2017-01-25 08:04:00    DPR        11210   
156134     2016-09-01 15:57:00 2017-01-25 08:05:00    DPR        11234   
156523     2016-09-01 09:04:00 2017-01-27 14:12:00    DOT        11357   
157320     2016-09-01 12:49:00 2017-01-23 12:00:00    DPR        11205   
158319     2016-09-01 17:07:00 2017-01-23 00:00:00    DOB        11223   
168301     2016-09-05 14:26:00 2017-01-27 10:26:00    DPR        11354   
172151     2016-09-02 11:39:00 2017-01-25 00:00:00    DOB        11356   
172709     2016-09-02 11:40:00 2017-01-25 00:00:00    DOB        11354   
172761     2016-09-02 15:24:00 2017-01-23 00:00:00    DOB        11235   
173841     2016-09-02 11:38:00 2017-01-24 10:02:00    DPR        11228   
174256     2016-09-02 14:07:00 2017-01-28 00:00:00    DOB        11373   
194295     2016-09-02 15:33:00 2017-01-25 08:04:00    DPR        11234   
194296     2016-09-02 15:47:00 2017-01-24 00:00:00    DOB        11356   
223853     2016-09-03 16:22:00 2017-01-24 10:01:00    DPR        11223   
224063     2016-09-03 10:34:00 2017-01-25 00:00:00    DOB        11220   
279023     2016-09-07 14:05:00 2017-01-26 17:01:00    DPR        11225   
279371     2016-09-07 14:17:00 2017-01-26 17:01:00    DPR        11225   
279375     2016-09-07 12:13:00 2017-01-27 15:01:00    DPR        11226   
290214     2016-09-04 13:06:00 2017-01-27 08:36:00    DOT        11435   
299493     2016-09-02 14:08:00 2017-01-23 07:13:00    DOT        11385   
299505     2016-09-03 15:38:00 2017-01-26 16:02:00    DPR        10458   
303524     2016-09-01 04:13:00 2017-01-22 00:00:00    DOB        11413   
338097     2016-09-01 13:26:00 2017-01-24 00:00:00    DOB        11379   
338103     2016-09-03 10:42:00 2017-01-24 00:00:00    DOB        11220   
338107     2016-09-07 13:49:00 2017-01-28 14:01:00    DPR        11210   
370926     2016-09-05 09:22:00 2017-01-27 08:59:00    DPR        11366   
431150     2016-09-05 00:46:00 2017-01-23 11:16:00    DOT        11233   
470712     2016-09-03 14:08:00 2017-01-25 00:00:00    DOB        11354   
472657     2016-09-02 08:46:00 2017-01-24 18:00:00    DPR        11238   
472768     2016-09-02 17:08:00 2017-01-27 08:01:00    DPR        11214   
473340     2016-09-02 14:35:00 2017-01-27 08:02:00    DPR        11223   
473358     2016-09-02 17:07:00 2017-01-27 08:02:00    DPR        11214   
473549     2016-09-02 09:31:00 2017-01-25 08:01:00    DPR        11226   
477723     2016-09-06 16:16:00 2017-01-28 14:02:00    DPR        11210   
478807     2016-09-06 11:59:00 2017-01-28 15:01:00    DPR        11103   
478939     2016-09-06 17:30:00 2017-01-27 00:00:00    DOB        11235   
521571     2016-09-07 00:55:00 2017-01-25 08:01:00    DPR        11234   
522055     2016-09-06 09:42:00 2017-01-27 00:00:00    DOB        11235   
533600     2016-09-07 19:36:00 2017-01-26 10:16:00    DOT        11412   
558640     2016-09-08 13:36:00 2017-01-29 14:00:00    DPR        11370   
569755     2016-09-08 09:48:00 2017-01-28 15:01:00    DPR        11204   
614925     2016-09-06 15:30:00 2017-01-25 16:22:00    DPR        11412   
681064     2016-09-01 23:31:00 2017-01-24 14:11:00    DOT        10469   
711401     2016-09-07 09:57:00 2017-01-27 00:00:00    DOB        11373   
836666     2016-09-10 12:25:00 2017-01-29 00:00:00    DOB        11423   

             Borough   Latitude  Longitude   processing_time  
Unique Key                                                    
5371          QUEENS  40.770843 -73.904511 

<h3>Looks like the upper end makes sense but the negative times don't</h3>
<h3>Though we need to explore this more, we'll get rid of negative times for now</h3>

In [35]:
data = data[data['processing_time']>=datetime.timedelta(0,0,0)]

In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 799325 entries, 11 to 971063
Data columns (total 8 columns):
Created Date       799325 non-null datetime64[ns]
Closed Date        799325 non-null datetime64[ns]
Agency             799325 non-null object
Incident Zip       799325 non-null object
Borough            799325 non-null object
Latitude           799325 non-null float64
Longitude          799325 non-null float64
processing_time    799325 non-null timedelta64[ns]
dtypes: datetime64[ns](2), float64(2), object(3), timedelta64[ns](1)
memory usage: 54.9+ MB


<h2>Finally, let's write a function that incorporates all our changes</h2>

In [ ]:
def read_311_data(datafile):
    import pandas as pd
    import numpy as np
    #Add the fix_zip function
    def fix_zip(input_zip):
        try:
            input_zip = int(float(input_zip))
        except:
            try:
                input_zip = int(input_zip.split('-')[0])
            except:
                return np.NaN
        if input_zip < 10000 or input_zip > 19999:
            return np.NaN
        return str(input_zip)
    
    #Read the file
    df = pd.read_csv(datafile,index_col='Unique Key')
    
    #fix the zip
    df['Incident Zip'] = df['Incident Zip'].apply(fix_zip)
    
    #drop all rows that have any nans in them (note the easier syntax!)
    
    df = df.dropna(how='any')
    
    #get rid of unspecified boroughs
    df = df[df['Borough'] != 'Unspecified']
    
    #Convert times to datetime and create a processing time column
    
    import datetime
    df['Created Date'] = df['Created Date'].apply(lambda x:datetime.datetime.strptime(x,'%m/%d/%Y %I:%M:%S %p'))
    df['Closed Date'] = df['Closed Date'].apply(lambda x:datetime.datetime.strptime(x,'%m/%d/%Y %I:%M:%S %p'))
    df['processing_time'] =  df['Closed Date'] - df['Created Date']
    
    #Finally, get rid of negative processing times and return the final data frame
    
    df = df[df['processing_time']>=datetime.timedelta(0,0,0)]
    
    return df
    

In [ ]:
df = read_311_data('nyc_311_data_subset.csv')
df.info()